### PR TITLE
Add events for debuggable apps

### PR DESCRIPTION
**Pull request by TsvetanMilanov**
Add `debuggableAppFound` and `debuggableAppLost` events which are raised when an application becomes available for debugging or when an application is not available for debugging anymore.

### CODE REVIEW

**Comment by justcodebuilduser** [comment]
:heart:

File: mobile/ios/device/ios-application-manager.ts [hunk 1]
```diff
@@ -202,11 +204,16 @@ export class IOSApplicationManager extends ApplicationManagerBase {
 		return this.$hostInfo.isDarwin || (this.$hostInfo.isWindows && this.$staticConfig.enableDeviceRunCommandOnWindows);
 	}
 
+	public getDebuggableApps(): IFuture<Mobile.IAndroidApplicationInformation[]> {
+		// Implement when we can fin debuggable applications for iOS.
```
Comment (rosen-vladimirov) [inline mobile/ios/device/ios-application-manager.ts]: fin -> find

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: README.md [hunk 2]
```diff
@@ -488,26 +527,33 @@ interface IAndroidApplicationInformation {
 
 Sample usage:
 ```JavaScript
-require("mobile-cli-lib").devicesService.getApplicationsAvailableForDebugging("4df18f307d8a8f1b")
-	.then(function(applicationsInfo) {
-		applicationsInfo.forEach(function(applicationInfo) {
-			console.log(applicationInfo);
+Promise.all(require("mobile-cli-lib").devicesService.getApplicationsAvailableForDebugging(["4df18f307d8a8f1b"]))
```
Comment (rosen-vladimirov) [inline README.md]: `getDebuggableApps` maybe

File: mobile/mobile-core/devices-service.ts [hunk 3]
```diff
@@ -411,8 +411,15 @@ export class DevicesService implements Mobile.IDevicesService {
...
-		return this.$androidProcessService.getDebuggableApps(deviceIdentifier);
+	public getDebuggableApps(deviceIdentifiers: string[]): IFuture<Mobile.IAndroidApplicationInformation[]>[] {
+		return _.map(deviceIdentifiers, (deviceIdentifier: string) => this.getProcessService(deviceIdentifier).getDebuggableApps(deviceIdentifier));
+	}
+
+	private getProcessService(deviceIdentifier: string): { getDebuggableApps: (deviceIdentifier: string) => IFuture<Mobile.IAndroidApplicationInformation[]> } {
+		let device = this.getDeviceByIdentifier(deviceIdentifier);
+		let iosProcessService = { getDebuggableApps: () => Future.fromResult(<Mobile.IAndroidApplicationInformation[]>[]) };
+
+		return device.deviceInfo.platform.toLocaleLowerCase() === this.$devicePlatformsConstants.Android.toLocaleLowerCase() ? this.$androidProcessService : iosProcessService;
```
Comment (rosen-vladimirov) [inline mobile/mobile-core/devices-service.ts]: toLocaleLowerCase -> toLowerCase

File: mobile/mobile-core/devices-service.ts [hunk 4]
```diff
@@ -411,8 +411,15 @@ export class DevicesService implements Mobile.IDevicesService {
 	}
 
 	@exportedPromise("devicesService")
-	public getDebuggableApps(deviceIdentifier: string): IFuture<Mobile.IAndroidApplicationInformation[]> {
-		return this.$androidProcessService.getDebuggableApps(deviceIdentifier);
+	public getDebuggableApps(deviceIdentifiers: string[]): IFuture<Mobile.IAndroidApplicationInformation[]>[] {
+		return _.map(deviceIdentifiers, (deviceIdentifier: string) => this.getProcessService(deviceIdentifier).getDebuggableApps(deviceIdentifier));
+	}
+
+	private getProcessService(deviceIdentifier: string): { getDebuggableApps: (deviceIdentifier: string) => IFuture<Mobile.IAndroidApplicationInformation[]> } {
```
Comment (rosen-vladimirov) [inline mobile/mobile-core/devices-service.ts]: I suggest skipping this check and use the `getDebuggableApps` from device's applicationManager:
```TypeScript
@exportedPromise("devicesService")
	public getDebuggableApps(deviceIdentifiers: string[]): IFuture<Mobile.IAndroidApplicationInformation[]>[] {
		return _.map(deviceIdentifiers, (deviceIdentifier: string) => this.getDebuggableAppsCore(deviceIdentifier));
	}

	private getDebuggableAppsCore(deviceIdentifier): IFuture<Mobile.IAndroidApplicationInformation[]> {
		return ((): Mobile.IAndroidApplicationInformation[] => {
			let device = this.getDeviceByIdentifier(deviceIdentifier);
			return device.applicationManager.getDebuggableApps().wait();
		}).future<Mobile.IAndroidApplicationInformation[]>()();
	}
```

**Comment by justcodebuilduser** [comment]
:heart:

File: README.md [hunk 5]
```diff
@@ -488,26 +527,33 @@ interface IAndroidApplicationInformation {
 
 Sample usage:
 ```JavaScript
-require("mobile-cli-lib").devicesService.getApplicationsAvailableForDebugging("4df18f307d8a8f1b")
-	.then(function(applicationsInfo) {
-		applicationsInfo.forEach(function(applicationInfo) {
-			console.log(applicationInfo);
+Promise.all(require("mobile-cli-lib").devicesService.getDebuggableApps(["4df18f307d8a8f1b"]))
```
Comment (rosen-vladimirov) [inline README.md]: add `JJY5KBTW75TCHQUK` to the array passed to `getDebuggableApps` as the sample result contains results for this identifier as well.

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by rosen-vladimirov** [comment]
:+1:

**Pull request merged**